workload/kv: additional parameters for workload generation

aliher1911 opened the pull request:
This PR adds following features:

This commit adds --sfu-wait-delay option to configure delay between 'select for update' and upsert statements. Tidy up of RuntimeOnly flags to avoid fixture from exposing timeouts etc for data generations options.

Refactors validation and configuration of generator operations and extracts validated parameters to generator and sequence for later use.

Adds --key-size flag to allow creation of primary keys of arbitrary sizes. When key size is used, strings would be used as keys. Key is prefixed by uint64 zero padded string to maintain sort order and allow pre splitting ranges same way as int keys. The rest of the key is filled with random characters with generator seeded with numeric part of the key. That allows us to keep keys less compressible than constants and at the same time allow updates if key cycle is short.

Adds support for string key imports in fixtures to speed up data load.

Epic: none

Release note: None

aliher1911 committed:
workload/kv: configurable delay for select for update

This commit adds --sfu-wait-delay option to configure delay
between 'select for update' and upsert statements.

Epic: none

Release note: None

aliher1911 committed:
workload/kv: refactor key sequence and generator

This commit separates parameter validation and creation of generator
and sequence for keys to avoid multiple similar checks thoughout code.

Epic: none

Release note: None

aliher1911 committed:
workload/kv: add key length flag

This commit adds --key-size flag which changes primary key to become
a string of specified length. String is prefixed with int
representation of key followed by consistent random sequence.

Epic: none

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100396)
<!-- Reviewable:end -->

kvoli commented:
Is this good to look at now @aliher1911? It is still a draft atm.

aliher1911 commented:
@kvoli ignore it for now. i figured i got some things wrong and realized why validate can't change state. In some places workload tables are used without parsing flags. 😞 
I'll fix it once I make build work again.

aliher1911 commented:
Sanity checking that generators are not broken and sharded schema is fixed.

# Sequential
```
$ ./cockroach workload init kv --drop --num-shards 5 --enum --secondary-index --sequential --key-size 25 --sequential
$ ./cockroach workload run kv --num-shards 5 --enum --secondary-index --key-size 25 --sequential --concurrency 1 --max-rate 1 --duration 10s
```
Resulting table:
```
root@localhost:26257/kv> select * from kv order by k;
              k             |  v   | e
----------------------------+------+----
  00000000000000000000mUNER | \x2b | v
  00000000000000000001BpLnf | \x6b | v
  00000000000000000002KSiOW | \xf9 | v
  00000000000000000003EjKsR | \xea | v
  00000000000000000004rgltB | \xd3 | v
  00000000000000000005igvMT | \xdb | v
  00000000000000000006OfAeI | \x0b | v
  00000000000000000007su1Z6 | \xf8 | v
  00000000000000000008AorCM | \x60 | v
  000000000000000000099Qc7d | \xb7 | v
  00000000000000000010wSv9w | \x88 | v
```

# Zipfian
```
$ ./cockroach workload init kv --drop --num-shards 5 --enum --secondary-index --key-size 25 --zipfian
$ ./cockroach workload run kv --num-shards 5 --enum --secondary-index --key-size 25 --zipfian --concurrency 1 --max-rate 1 --duration 10s
```
Resulting table:
```
root@localhost:26257/kv> select * from kv order by k;
              k             |  v   | e
----------------------------+------+----
  00000000000000000000mUNER | \xef | v
  00000000000000000004rgltB | \xea | v
  00000000000000000013enlVw | \xde | v
  00000000000000000078PLOEr | \x4c | v
  00000000000000001409d2FEb | \xd7 | v
  00000000000000012885Evaux | \x24 | v
  00000000000000530238ts3gz | \x20 | v
  00000000000018182288YKAsz | \xb1 | v
  00663765391359604224ZUi0I | \x9e | v
```

# Hash
```
$ ./cockroach workload init kv --drop --enum --secondary-index --key-size 25
$ ./cockroach workload run kv --enum --secondary-index --key-size 25 --concurrency 1 --max-rate 1 --duration 10s
```
Resulting table:
```
root@localhost:26257/kv> select * from kv order by k;
              k             |  v   | e
----------------------------+------+----
  01106988831625274679zghIY | \xca | v
  02732189456399973006jDux9 | \xf5 | v
  053635977682905575248PIcE | \x5a | v
  08293266398741977489jzHpM | \xd7 | v
  10299586364998161683rslnR | \x4c | v
  10879287192010464812BkRur | \xf2 | v
  12393558243119987931Anytn | \x65 | v
  12605989761922436892aWLY1 | \xf7 | v
  15004972025794792726S8Vwc | \xac | v
  18445512827218951566ilHIq | \x37 | v
```

Random with int keys:
```
$ ./cockroach workload init kv --drop --secondary-index
$ ./cockroach workload run kv --secondary-index --concurrency 1 --max-rate 1 --duration 10s
```
Resulting table:
```
root@localhost:26257/kv> select * from kv order by k;
           k           |  v
-----------------------+-------
  -8773477328541093751 | \xf5
  -4654277624974166269 | \x6c
  -2454105589580113183 | \x29
  -2110775707891572897 | \x62
    343859946065825060 | \x82
   1395394053388110161 | \xac
   1415862316324127089 | \x07
   1603802700825534582 | \xd7
   3962375795458359291 | \xb6
   5191688387284494824 | \xc8
   7861581158979270000 | \x75
```

aliher1911 commented:
@kvoli should be good to look at. I also found that shards+splits doesn't work with new string keys (fails in SQL), but that's probably ok for now as for int keys shard+splits work but not as intended (will silently have incorrect split points).

aliher1911 committed:
workload/kv: add support for import for string keys

Previously --key-size was not compatible with --insert-count.
This commit adds batch add support for string keys to improve
workload prepare performance.

Epic: none

Release note: None

aliher1911 commented:
FYI: if you already had a look, I also added import for long keys to enable fixtures and tidied up fixture flags. Previously we didn't mark them as `RuntimeOnly` so workloads with different read percentage or timeouts were technically different fixtures while having identical data.

kvoli reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice cleanup and new features.

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-:-NSkkSElB8_1X_I-x0c5:bzdcu2)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aliher1911, @herkolategan, and @smg260)

___
*[`pkg/workload/kv/kv.go` line 327 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-NSkjJgQ0VsqHOp03R2X:-NSkjJgQ0VsqHOp03R2Y:bbsfgv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/220d25ad425a30f83c8c117a5a3d670824e9314b/pkg/workload/kv/kv.go#L327)):*
> ```Go
> func splitFinder(i, splits int, r keyRange, k keyTransformer) interface{} {
> 	if i >= splits {
> 		panic(fmt.Sprintf("programming error: split index (%d) cannot be less than 0, "+
> ```

nit: The 0 check was removed (and redundant due to flag parsing). Should we remove the 0 reference here too.
___
*[`pkg/workload/kv/kv.go` line 713 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-NSkhA1c4EdAqYPyTbXF:-NSkhA1c4EdAqYPyTbXG:b-nigrry) ([raw file](https://github.com/cockroachdb/cockroach/blob/220d25ad425a30f83c8c117a5a3d670824e9314b/pkg/workload/kv/kv.go#L713)):*
> ```Go
> // mapping integer key into string of arbitrary size for testing keys size.
> // First two methods are used in table creation and ops, while last ones are
> // only needed to for import. If transformer doesn't support import, it could
> ```

nit: only need for import.
___
*[`pkg/workload/kv/kv.go` line 754 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-NSkk30v3AmRYd9j5EoD:-NSkk30v3AmRYd9j5EoE:blx2k8e) ([raw file](https://github.com/cockroachdb/cockroach/blob/220d25ad425a30f83c8c117a5a3d670824e9314b/pkg/workload/kv/kv.go#L754)):*
> ```Go
> }
> 
> const minStringKeyDigits = 20
> ```

How was 20 selected?


<!-- Sent from Reviewable.io -->

aliher1911 reviewed (commented):
Thanks for the read!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-:-NSlkEW53YvgRi2S7cHy:b55vvf7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @herkolategan, @kvoli, and @smg260)

___
*[`pkg/workload/kv/kv.go` line 327 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-NSkjJgQ0VsqHOp03R2X:-NSlk-_P8UuOzZjt4RKM:bjlqm1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/220d25ad425a30f83c8c117a5a3d670824e9314b/pkg/workload/kv/kv.go#L327)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

nit: The 0 check was removed (and redundant due to flag parsing). Should we remove the 0 reference here too.
</blockquote></details>

I returned the 0 check. So defensive from kv to mistrust whoever invokes imports. :laughing: 
___
*[`pkg/workload/kv/kv.go` line 754 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/100396#-NSkk30v3AmRYd9j5EoD:-NSkkqOc2lYDKommX9M9:b-tn6y7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/220d25ad425a30f83c8c117a5a3d670824e9314b/pkg/workload/kv/kv.go#L754)):*
<details><summary><i>Previously, kvoli (Austen) wrote…</i></summary><blockquote>

How was 20 selected?
</blockquote></details>

20 chars needed to fit 64 bit unsigned int. Technically we can do maths and look on max range to figure out min length, but I think it is not worth the complexity. I put an explaining comment.


<!-- Sent from Reviewable.io -->

aliher1911 commented:
bors r=kvoli

aliher1911 commented:
bors r=kvoli

aliher1911 commented:
bors r=kvoli
